domains/client/domain-executor/src/domain_block_preprocessor.rs

```diff
@@ -0,0 +1,300 @@
+        .runtime_api()
+        .extrinsics_shuffling_seed(block_hash, header)?;
+
+    let domain_bundles = if domain_id.is_system() {
+        let (system_bundles, core_bundles) = primary_chain_client
+            .runtime_api()
+            .extract_system_bundles(block_hash, extrinsics)?;
+        DomainBundles::System(system_bundles, core_bundles)
+    } else if domain_id.is_core() {
+        let core_bundles = primary_chain_client
+            .runtime_api()
+            .extract_core_bundles(block_hash, extrinsics, domain_id)?;
+        DomainBundles::Core(core_bundles)
+    } else {
+        unreachable!("Open domains are unsupported")
+    };
+
+    Ok((domain_bundles, shuffling_seed, maybe_new_runtime))
+}
+
+pub(crate) fn compile_own_domain_bundles<Block, PBlock>(
+    bundles: OpaqueBundles<PBlock, Block::Hash>,
+) -> Vec<Block::Extrinsic>
+where
+    Block: BlockT,
+    PBlock: BlockT,
+{
+    bundles
+            .into_iter()
+            .flat_map(|bundle| {
+                bundle.extrinsics.into_iter().filter_map(|opaque_extrinsic| {
+                    match <<Block as BlockT>::Extrinsic>::decode(
+                        &mut opaque_extrinsic.encode().as_slice(),
+                    ) {
+                        Ok(uxt) => Some(uxt),
+                        Err(e) => {
+                            tracing::error!(
+                                error = ?e,
+                                "Failed to decode the opaque extrisic in bundle, this should not happen"
+                            );
+                            None
+                        },
+                    }
+                })
+            })
+            .collect::<Vec<_>>()
+}
+
+pub(crate) fn deduplicate_and_shuffle_extrinsics<Block, Client>(
+    client: &Arc<Client>,
+    parent_hash: Block::Hash,
+    mut extrinsics: Vec<Block::Extrinsic>,
+    shuffling_seed: Randomness,
+) -> Result<Vec<Block::Extrinsic>, sp_blockchain::Error>
+where
+    Block: BlockT,
+    Client: ProvideRuntimeApi<Block>,
+    Client::Api: DomainCoreApi<Block, AccountId>,
+{
+    let mut seen = Vec::new();
+    extrinsics.retain(|uxt| match seen.contains(uxt) {
+        true => {
+            tracing::trace!(extrinsic = ?uxt, "Duplicated extrinsic");
+            false
+        }
+        false => {
+            seen.push(uxt.clone());
+            true
+        }
+    });
+    drop(seen);
+
+    tracing::trace!(?extrinsics, "Origin deduplicated extrinsics");
+
+    let extrinsics: Vec<_> = match client.runtime_api().extract_signer(parent_hash, extrinsics) {
+        Ok(res) => res,
+        Err(e) => {
+            tracing::error!(error = ?e, "Error at calling runtime api: extract_signer");
+            return Err(e.into());
+        }
+    };
+
+    let extrinsics = shuffle_extrinsics::<<Block as BlockT>::Extrinsic>(extrinsics, shuffling_seed);
+
+    Ok(extrinsics)
+}
+
+pub struct SystemDomainBlockPreprocessor<Block, PBlock, Client, PClient> {
```

Comment:
maybe move it under `system_bundle_preprocessor` since we the similar structure for processor as well for consistency

Reply:
It's not the final form, some further refactoring changes are expected in the follow up PR so I kept it as a single file for simplicity.